商品購入機能

tech-o-zukky opened the pull request:
レビューよろしくお願いします。

[単体テスト](https://gyazo.com/d2f764a083658fa638f7e827eb10a73c)

# 動画

[一覧->詳細->購入ページ。選択した商品が表示されていること](https://gyazo.com/148589438a5208f298caa82408a316a5)

[ログアウト状態のユーザーは、URLを直接入力して商品購入ページに遷移しようとすると、商品の販売状況に関わらずログインページに遷移](https://gyazo.com/faa7e0d2b264bd3d0872afcdcad45164)

[ログイン状態の出品者が、URLを直接入力して自身の出品した商品購入ページに遷移しようとすると、トップページに遷移すること](https://gyazo.com/fe430380765538ca24b1752a1c93eee0)

[ログイン状態のユーザーが、URLを直接入力して売却済み商品の商品購入ページへ遷移しようとすると、トップページに遷移すること](https://gyazo.com/2dae800fb3dc50db602a2766c8c21745)

[入力に問題がある状態で購入ボタンが押されたら、購入ページに戻りエラーメッセージが表示されること](https://gyazo.com/1c1e756a6cccab9280b88d630ac21f37)

tech-o-zukky committed:
orderコントロール生成

tech-o-zukky committed:
DLしたビューとCSS配置

tech-o-zukky committed:
ルーティング設定：商品購入

・購入ページ初期表示(index)
・購入処理(create)

tech-o-zukky committed:
orderコントローラにアクション追加

・ログアウト済みユーザーが遷移したら、ログインページへ遷移

tech-o-zukky committed:
商品購入ページ：初期表示の実装

tech-o-zukky committed:
purchaseモデル生成

tech-o-zukky committed:
addressモデル生成

tech-o-zukky committed:
db:migrate実行

purchaseとaddressのmigrate実行後にcommit

tech-o-zukky committed:
usersにpurchasesのアソシエーション追加

・itemsとのアソシエーションも漏れてたので追加

tech-o-zukky committed:
itemsにpurchaseとのアソシエーション追加

tech-o-zukky committed:
purchaseにアソシエーション追加

tech-o-zukky committed:
addressにアソシエーション追加

tech-o-zukky committed:
prefectureにアソシエーション追加

tech-o-zukky committed:
詳細画面から購入画面へのパス追加

tech-o-zukky committed:
購入フォーム実装

tech-o-zukky committed:
購入フォーム実装（ビュー）

tech-o-zukky committed:
turbolinksの記述を除外

tech-o-zukky committed:
payjp.js読み込み

tech-o-zukky committed:
カード情報トークン化用JS作成と読み込み

tech-o-zukky committed:
GEM:PAY.JPインストール

tech-o-zukky committed:
カード決済JSの実装

※公開鍵はあとで環境変数化

tech-o-zukky committed:
コントローラー：購入時決済実装

PAY.jp用のキーは別途設定

tech-o-zukky committed:
購入時Formオブジェクトパターン実装

tech-o-zukky committed:
決済処理：環境変数化

tech-o-zukky committed:
購入機能：URL直接入力チェック実装

tech-o-zukky committed:
商品が購入済みだったときの条件分岐

tech-o-zukky committed:
購入画面：エラーメッセージの処理

tech-o-zukky committed:
購入用Formオブジェクト:単体テストファイル生成

tech-o-zukky committed:
購入処理:不要な記述を削除

tech-o-zukky committed:
電話番号：バリデーション修正

10〜11桁から11桁以下に修正

tech-o-zukky committed:
RSpec:購入用Form単体テスト実装

tech-o-zukky committed:
Rubocop実行後

tech-o-zukky committed:
URL直接入力チェック修正

harukitoda commented on spec/models/order_information_spec.rb:
```diff
@@ -0,0 +1,121 @@
+      end
+    end
+  end
+end
```
テストコードに以下の項目を追加するようにしましょう！
・prefecture_id が０以外でないと登録できないこと
・数字のみでないと登録できないこと

harukitoda reviewed (commented):
レビュー依頼ありがとうございます！
1点コメントしましたのでご確認お願い致します！

tech-o-zukky commented:
レビューありがとうございます。
指摘を受けて、テストケース追加しました（計3件）

- prefecture_id が０以外でないと登録できないこと
都道府県:0の場合はエラー

- 数字のみでないと登録できないこと
郵便番号:フォーマットが合わない場合はエラー(ハイフンと数字。123-4567となる)
電話番号:数字以外はエラーであること(09012345678が正))

[単体テスト2回目](https://gyazo.com/32658a13b70c2bb0cc5c951b3f893609)

tech-o-zukky committed:
レビュー後：テストケース追加

- prefecture_id が０以外でないと登録できないこと
都道府県:0の場合はエラー

- 数字のみでないと登録できないこと
郵便番号:フォーマットが合わない場合はエラー(ハイフンと数字。123-4567となる)
電話番号:数字以外はエラーであること(09012345678が正))

iy06 commented on app/models/order_information.rb:
```diff
@@ -0,0 +1,25 @@
+class OrderInformation
+  include ActiveModel::Model
+
+  attr_accessor :token, :card_no, :card_exp_month, :card_exp_year, :card_cvc,
```
4行目の:card_no, :card_exp_month, :card_exp_year, :card_cvc,ですが、カード情報は保存しないため削除しましょう。

iy06 commented on app/controllers/orders_controller.rb:
```diff
@@ -0,0 +1,44 @@
+  end
+
+  def create
+    @item = Item.find(params[:item_id])
```
`@item = Item.find(params[:item_id])`こちらの記述が二つ存在しておりますのでDRYに則りメソッド化しましょう！

メソッド化した後、必要なアクションやメソッドにbefore_actionを使って呼び出しましょう。

iy06 reviewed (commented):
レビュー依頼ありがとうございます！

修正箇所をコメントさせて頂きましたのでご確認お願いします。

tech-o-zukky commented:
ご指摘ありがとうございます。
肝心なところを削除漏れしていました。

修正しましたので、再レビューお願いします。
以下は再取得した動画です。

[修正後・詳細→購入画面への遷移](https://gyazo.com/a2ca94f29e69dbeb43d0691fcf8a2f16)

[(続き)修正後・購入画面→(購入処理:create)→トップ画面への遷移](https://gyazo.com/cab7cd403636c890b96a018989d2bc0b)

tech-o-zukky committed:
レビュー後修正

・一部記述削除
・処理のまとめ

iy06 reviewed (commented):
素晴らしいPRありがとうございます！

LGTMとさせていただきます！実装お疲れ様でした！
引き続き実装頑張っていきましょう！
